File: trunk/theLink/example/perl/Filter4.pl
#+
#§  \file       theLink/example/perl/Filter4.pl
#§  \brief      \$Id: Filter4.pl 92 2009-12-21 11:58:07Z aotto1968 $
#§  
#§  (C) 2009 - NHI - #1 - Project - Group
#§  
#§  \version    \$Rev: 92 $
#§  \author     EMail: aotto1968 at users.berlios.de
#§  \attention  this software has GPL permissions to copy
#§              please contact AUTHORS for additional information
#§

use strict;
use Net::PerlMsgque;
use FileHandle;

$| = 1;

package Filter4;
use base qw(Net::PerlMsgque::MqS);

  sub ErrorWrite {
    my $ctx = shift;
    my $FH = $ctx->DictGet("FH");
    if (defined($FH)) {
      print $FH "ERROR: " . $ctx->ErrorGetText() . "\n";
    } else {
      $ctx->Log("ErrorWrite", 0, $ctx->ErrorGetText() . "\n");
    }
    $ctx->ErrorReset();
  }

  sub WRIT {
    my $ftr = shift;
    my $FH = $ftr->ServiceGetFilter()->DictGet("FH");
    if (defined($FH)) {
      print $FH $ftr->ReadC() . "\n";
    } else {
      $ftr->Log("ErrorWrite", 0, $ftr->ReadC() . "\n");
    }
    $ftr->SendRETURN();
  }

  sub EXIT {
    exit 1;
  }

  sub LOGF {
    my $ctx = shift;
    my $ftr = $ctx->ServiceGetFilter();
    if ($ftr->LinkGetTargetIdent() eq "transFilter") {
      $ctx->ReadForward($ftr);
    } else {
      open(my $FH, ">>", $ctx->ReadC());
      $FH->autoflush(1);
      $ctx->DictSet("FH", $FH);
    }
    $ctx->SendRETURN();
  }

  sub Event {
    my $ctx = shift;
    my $itms = $ctx->DictGet("itms");
    my $it = @{$itms}[0];
    if (!defined($it)) {
      $ctx->ErrorSetCONTINUE();
    } else {
      eval {
	my $ftr = $ctx->ServiceGetFilter();
	$ftr->LinkConnect();
	$ctx->ReadLOAD($it);
	$ctx->ReadForward($ftr);
      };
      if ($@) {
	$ctx->ErrorSet($@);
	if ($ctx->ErrorIsEXIT) {
	  $ctx->ErrorReset();
	  return;
	} else {
	  $ctx->ErrorWrite();
	}
      }
      shift(@{$itms});
    }
  }

  sub FilterIn {
    my $ctx = shift;
    my $itms = $ctx->DictGet("itms");
    push(@{$itms}, $ctx->ReadDUMP());
    $ctx->SendRETURN();
  }

  sub ServerCleanup {
    my $ctx = shift;
    my $FH = $ctx->DictGet("FH");
    if (defined($FH)) {
      close($FH);
    }
  }

  sub ServerSetup {
    my $ctx = shift;
    my $ftr = $ctx->ServiceGetFilter();
    $ctx->ServiceCreate("+ALL", \&FilterIn);
    $ctx->ServiceCreate("LOGF", \&LOGF);
    $ctx->ServiceCreate("EXIT", \&EXIT);
    $ftr->ServiceCreate("WRIT", \&WRIT);
    $ftr->ServiceProxy("+TRT");
    $ctx->DictSet("itms", []);
    $ctx->DictSet("FH", undef);

  }

  sub new {
    my $class = shift;
    my $ctx = $class->SUPER::new(@_);
    $ctx->ConfigSetIgnoreExit(1);
    $ctx->ConfigSetServerSetup(\&ServerSetup);
    $ctx->ConfigSetServerCleanup(\&ServerCleanup);
    $ctx->ConfigSetEvent(\&Event);
    bless $ctx, $class;
    return $ctx;
  }


package main;

  Net::PerlMsgque::MqFactoryS::Default("transFilter", "Filter4");
  our $srv = new Filter4();
  eval {
    $srv->LinkCreate(@ARGV);
    $srv->ProcessEvent(Net::PerlMsgque::WAIT_FOREVER);
  };
  if ($@) {
    $srv->ErrorSet($@);
  }
  $srv->Exit();

1;

File: trunk/theLink/example/go/Filter4.go
/**
 *  \file       theLink/example/go/Filter4.go
 *  \brief      \$Id$
 *  
 *  (C) 2009 - NHI - #1 - Project - Group
 *  
 *  \version    \$Rev$
 *  \author     EMail: aotto1968 at users.berlios.de
 *  \attention  this software has GPL permissions to copy
 *              please contact AUTHORS for additional information
 */

package main

import (
  . "gomsgque"
    "container/list"
    "os"
)

type Filter4 struct {
  *MqS
  itms	*list.List
  FH	*os.File
}

func NewFilter4(tmpl *MqS) *MqS {
  ret := new(Filter4)
  ret.MqS = NewMqS(tmpl, ret)
  ret.itms = list.New()
  return ret.MqS
}

func (this *Filter4) ErrorWrite() {
  this.FH.WriteString("ERROR: " + this.ErrorGetText() + "\n")
  this.ErrorReset()
}

type LOGF Filter4
  func (this *LOGF) Call() {
    ftr := this.ServiceGetFilter()
    if (ftr.LinkGetTargetIdent() == "transFilter") {
      this.ReadForward(ftr)
    } else {
      this.FH,_ = os.Open(this.ReadC(),os.O_WRONLY|os.O_APPEND,0666)
    }
    this.SendRETURN()
  }

type EXIT Filter4
  func (this *EXIT) Call() {
    os.Exit(1)
  }

type WRIT Filter4
  func (this *WRIT) Call() {
    this.ServiceGetFilter().GetSelf().(*Filter4).FH.WriteString(this.ReadC() + "\n")
    this.SendRETURN()
  }

type ALLS Filter4
  func (this *ALLS) Call() {
    this.itms.PushBack(this.ReadDUMP())
    this.SendRETURN()
  }

func (this *Filter4) Event() {
  if (this.itms.Len() <= 0) {
    this.ErrorSetCONTINUE()
  } else {
    it := this.itms.Front()
    defer func() {
      if x := recover(); x != nil {
	this.ErrorSet(x)
	if this.ErrorIsEXIT() {
	  this.ErrorReset()
	  return;
	} else {
	  this.ErrorWrite()
	}
      }
      this.itms.Remove(it)
    }()
    ftr := this.ServiceGetFilter()
    ftr.LinkConnect()
    this.ReadLOAD(it.Value.(*MqDumpS))
    this.ReadForward(ftr)
  }
}

func (this *Filter4) ServerCleanup() {
  if (this.FH != nil) {
    this.FH.Close()
    this.FH = nil
  }
}

func (this *Filter4) ServerSetup() {
  ftr := this.ServiceGetFilter().GetSelf().(*Filter4)
  this.ServiceCreate("LOGF", (*LOGF)(this))
  this.ServiceCreate("EXIT", (*EXIT)(this))
  this.ServiceCreate("+ALL", (*ALLS)(this))
  ftr.ServiceCreate("WRIT", (*WRIT)(ftr))
  ftr.ServiceProxy("+TRT")
}

func main() {
  FactoryDefault("transFilter", NewFilter4)
  srv := NewFilter4(nil)
  defer func() {
    if x := recover(); x != nil {
      srv.ErrorSet(x)
    }
    srv.Exit()
  }()
  srv.ConfigSetIgnoreExit(true)
  srv.ConfigSetName("Filter4")
  srv.LinkCreate(os.Args...)
  srv.ProcessEvent(WAIT_FOREVER)
}

